Add `core.txs`

mattromano opened the pull request:
# What's changed
1. Updated `txs.sql` following the schema design described and decided here #1 
2. Updated `schema.yml` txs model to include tests for newly added fields. Happy to update based on @chuxinh 's PR suggestion to split into two separate .yml files. 

# Testing

## dbt -run --select core.txs
``` 
root@209b99055b30:/harmony# dbt run --select core.txs
Running with dbt=0.21.1
Found 4 models, 29 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

15:10:33 | Concurrency: 4 threads (target='dev')
15:10:33 |
15:10:33 | 1 of 1 START incremental model DEV.txs............................... [RUN]
15:30:07 | 1 of 1 OK created incremental model DEV.txs.......................... [SUCCESS 1 in 1174.10s]
15:30:07 |
15:30:07 | Finished running 1 incremental model in 1177.01s.

Completed successfully 
```

## dbt -test

``` 
root@209b99055b30:/harmony# dbt test
Running with dbt=0.21.1
Found 4 models, 29 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

15:38:49 | Concurrency: 4 threads (target='dev')
15:38:49 | 
15:38:49 | 1 of 29 START test not_null_blocks_block_hash........................ [RUN]
15:38:49 | 2 of 29 START test not_null_blocks_block_id.......................... [RUN]
15:38:49 | 3 of 29 START test not_null_blocks_block_parent_hash................. [RUN]
15:38:49 | 4 of 29 START test not_null_blocks_block_timestamp................... [RUN]
15:38:52 | 4 of 29 PASS not_null_blocks_block_timestamp......................... [PASS in 3.45s]
15:38:52 | 5 of 29 START test not_null_blocks_gas_limit......................... [RUN]
15:38:52 | 2 of 29 PASS not_null_blocks_block_id................................ [PASS in 3.47s]
15:38:52 | 3 of 29 PASS not_null_blocks_block_parent_hash....................... [PASS in 3.47s]
15:38:52 | 1 of 29 PASS not_null_blocks_block_hash.............................. [PASS in 3.47s]
15:38:52 | 6 of 29 START test not_null_blocks_gas_used.......................... [RUN]
15:38:52 | 7 of 29 START test not_null_blocks_miner............................. [RUN]
15:38:52 | 8 of 29 START test not_null_metric_miner_count_metric_date........... [RUN]
15:38:54 | 7 of 29 PASS not_null_blocks_miner................................... [PASS in 1.92s]
15:38:54 | 9 of 29 START test not_null_metric_miner_count_metric_period......... [RUN]
15:38:55 | 6 of 29 PASS not_null_blocks_gas_used................................ [PASS in 2.15s]
15:38:55 | 10 of 29 START test not_null_metric_miner_count_miner_count.......... [RUN]
15:38:55 | 8 of 29 PASS not_null_metric_miner_count_metric_date................. [PASS in 2.15s]
15:38:55 | 11 of 29 START test not_null_txs_mr_block_hash....................... [RUN]
15:38:56 | 5 of 29 PASS not_null_blocks_gas_limit............................... [PASS in 3.30s]
15:38:56 | 12 of 29 START test not_null_txs_mr_block_number..................... [RUN]
15:38:56 | 9 of 29 PASS not_null_metric_miner_count_metric_period............... [PASS in 1.38s]
15:38:56 | 13 of 29 START test not_null_txs_mr_block_timestamp.................. [RUN]
15:38:56 | 10 of 29 PASS not_null_metric_miner_count_miner_count................ [PASS in 1.60s]
15:38:56 | 14 of 29 START test not_null_txs_mr_data............................. [RUN]
15:38:56 | 11 of 29 PASS not_null_txs_mr_block_hash............................. [PASS in 1.83s]
15:38:56 | 15 of 29 START test not_null_txs_mr_from_address..................... [RUN]
15:38:57 | 13 of 29 PASS not_null_txs_mr_block_timestamp........................ [PASS in 1.57s]
15:38:57 | 16 of 29 START test not_null_txs_mr_gas.............................. [RUN]
15:38:58 | 14 of 29 PASS not_null_txs_mr_data................................... [PASS in 1.37s]
15:38:58 | 17 of 29 START test not_null_txs_mr_gas_price........................ [RUN]
15:38:58 | 15 of 29 PASS not_null_txs_mr_from_address........................... [PASS in 1.25s]
15:38:58 | 18 of 29 START test not_null_txs_mr_hash............................. [RUN]
15:38:59 | 18 of 29 PASS not_null_txs_mr_hash................................... [PASS in 1.81s]
15:38:59 | 19 of 29 START test not_null_txs_mr_native_from_address.............. [RUN]
15:39:01 | 16 of 29 PASS not_null_txs_mr_gas.................................... [PASS in 3.37s]
15:39:01 | 20 of 29 START test not_null_txs_mr_native_to_address................ [RUN]
15:39:01 | 17 of 29 PASS not_null_txs_mr_gas_price.............................. [PASS in 3.00s]
15:39:01 | 21 of 29 START test not_null_txs_mr_nonce............................ [RUN]
15:39:01 | 19 of 29 PASS not_null_txs_mr_native_from_address.................... [PASS in 1.16s]
15:39:01 | 22 of 29 START test not_null_txs_mr_status........................... [RUN]
15:39:01 | 12 of 29 PASS not_null_txs_mr_block_number........................... [PASS in 5.07s]
15:39:01 | 23 of 29 START test not_null_txs_mr_to_address....................... [RUN]
15:39:03 | 21 of 29 PASS not_null_txs_mr_nonce.................................. [PASS in 2.16s]
15:39:03 | 24 of 29 START test not_null_txs_mr_value............................ [RUN]
15:39:03 | 20 of 29 PASS not_null_txs_mr_native_to_address...................... [PASS in 2.23s]
15:39:03 | 25 of 29 START test unique_blocks_block_hash......................... [RUN]
15:39:03 | 22 of 29 PASS not_null_txs_mr_status................................. [PASS in 2.20s]
15:39:03 | 26 of 29 START test unique_blocks_block_id........................... [RUN]
15:39:03 | 23 of 29 PASS not_null_txs_mr_to_address............................. [PASS in 2.30s]
15:39:03 | 27 of 29 START test unique_blocks_block_parent_hash.................. [RUN]
15:39:05 | 26 of 29 PASS unique_blocks_block_id................................. [PASS in 2.32s]
15:39:05 | 28 of 29 START test unique_blocks_block_timestamp.................... [RUN]
15:39:05 | 24 of 29 PASS not_null_txs_mr_value.................................. [PASS in 2.47s]
15:39:05 | 29 of 29 START test unique_txs_mr_hash............................... [RUN]
15:39:05 | 25 of 29 PASS unique_blocks_block_hash............................... [PASS in 2.54s]
15:39:06 | 27 of 29 PASS unique_blocks_block_parent_hash........................ [PASS in 2.70s]
15:39:07 | 28 of 29 PASS unique_blocks_block_timestamp.......................... [PASS in 1.47s]
15:39:28 | 29 of 29 PASS unique_txs_mr_hash..................................... [PASS in 23.00s]
15:39:28 | 
15:39:28 | Finished running 29 tests in 40.88s.

Completed successfully 
```

![Screenshot 2022-01-03 074453](https://user-images.githubusercontent.com/42412983/147950680-eaa5ea5b-757d-4097-b150-d8aa00f4e5e2.png)


Also, this is my first pull request ever so apologies if I am missing anything or committing any PR faux pas.

mattromano committed:
update txn sql

mattromano committed:
yaml update

mattromano committed:
add cluster and where clause

mattromano committed:
add dbt_docs to ignore

mattromano committed:
removed unique test from nonce and format

mattromano committed:
updated txs yaml and txs sql final

mattromano committed:
fixed schema.yml model name from txs_mr to txs

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,30 @@
---   "value": 0
--- }
\ No newline at end of file
+    tx :nonce :: string as nonce,
```
nit: you can remove the spacing here and the following

mattromano replied:
Spacing has been removed from the rest of the query

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,30 @@
---   "value": 0
--- }
\ No newline at end of file
+    tx :nonce :: string as nonce,
```
```suggestion
    tx:nonce::string as nonce,
```

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,30 @@
+    tx :gas as gas,
+    tx_id as "HASH",
+    tx :input :: string as "DATA",
+    case
+        when tx :receipt :status :: string = '0x0' then FALSE
+        else TRUE
+    end as "STATUS"
```
can simplify as the following
```suggestion
    tx:receipt:status::string = '0x1'  as "STATUS"
```

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,30 @@
+        when tx :receipt :status :: string = '0x0' then FALSE
+        else TRUE
+    end as "STATUS"
+from
+    {{ deduped_txs("harmony_txs") }}
```
```suggestion
from {{ deduped_txs("harmony_txs") }}
```

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,30 @@
+    end as "STATUS"
+from
+    {{ deduped_txs("harmony_txs") }}
+where
+    {{ incremental_load_filter("block_timestamp") }}
```
```suggestion
where {{ incremental_load_filter("block_timestamp") }}
```

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,30 @@
--- }
\ No newline at end of file
+    tx :nonce :: string as nonce,
+    tx_block_index as "INDEX",
```
totally understand why you're using the uppercase here but think snowflake is also accepting lower cases as all the columns in snowflake is auto uppercased :D

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: index
+        description: "Integer of the transactions index position in the block. null when its pending"
```
```suggestion
        description: "Integer of the transactions index position in the block. `null` when its pending".
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+        description: "Integer of the transactions index position in the block. null when its pending"
+
+      - name: native_from_address
+        description: "Native Harmony address of the sender"
```
```suggestion
        description: "Native Harmony address of the sender."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: native_to_address
+        description: "Native Harmony address of the reciever"
```
```suggestion
        description: "Native Harmony address of the receiver."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: from_address
+        description: "The Ethereum address of the sender"
```
```suggestion
        description: "The Ethereum address of the sender."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: to_address
+        description: "The Ethereum address of the reciever"
```
```suggestion
        description: "The Ethereum address of the receiver."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+          
+      - name: value
+        description: "The value transfered in ATTO format (a factor of 10^18)"
```
```suggestion
        description: "The value transfered in ATTO format (a factor of 10^18)."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: block_number
+        description: "Block number where this transaction was in. null when its pending."
```
```suggestion
        description: "Block number where this transaction was in. `null` when it's pending."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: block_hash
+        description: "Hash of the block where this transaction was in. null when its pending."
```
```suggestion
        description: "Hash of the block where this transaction was in. `null` when it's pending."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: hash
+        description: "The hash of the transaction, this is the primary key for this table"
```
```suggestion
        description: "The hash of the transaction, this is the primary key for this table."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
           - unique
+
+      - name: data
+        description: "Bytecode"
```
```suggestion
        description: "Bytecode of the data sent along with the transaction."
```

chuxinh commented on models/core/schema.yml:
```diff
@@ -10,45 +10,122 @@ models:
+          - not_null
+
+      - name: status
+        description: "True or False"
```
```suggestion
        description: "Boolean representing whether the transaction is successful or not."
```

chuxinh reviewed (changes requested):
table schema looks good to me. just some feedback for formats and comments.
now i'm seriously thinking if we should breakdown `schema.yml` into separate ones given multiple people working/will be working on this and the rebases we need to make as well 😅
cc: @drakedanner @jfmyers if you have any thought

mattromano committed:
removed spacing

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
simplified Status T/F

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
update dedupe macro format

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
update incremental filter format

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
index desc format/punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
block num format and punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
block hash format and punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
hash punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
bytecode desc

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
status desc update

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
spacing and header lowercase format

mattromano committed:
native_from desc punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
native_to desc update

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
from address punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
to_address punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
value desc punctuation

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
split .yml into txs.yml

mattromano committed:
added tx_hash header name -tentative

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,25 @@
-{{ config(materialized='incremental', unique_key='tx_id', tags=['core']) }}
+{{ config(
+    materialized = 'incremental',
+    unique_key = 'HASH',
```
```suggestion
    unique_key = 'tx_hash',
```

chuxinh commented on models/core/txs.yml:
```diff
@@ -0,0 +1,82 @@
+models:
+
+  - name: txs
+    description: "Harmony Transactions"
```
```suggestion
    description:  |-
      Harmony Transactions.
```

chuxinh commented on models/core/txs.yml:
```diff
@@ -0,0 +1,82 @@
+    description: "Harmony Transactions"
+    columns:
+      - name: block_timestamp
+        description: "The time when the block was mined."
```
Actually can leave out the quotation marks here and the following

chuxinh commented on models/core/txs.yml:
```diff
@@ -0,0 +1,82 @@
+
+      - name: block_hash
+        description: "Hash of the block where this transaction was in. `null` when it's pending."
+        tests:
+          - not_null
```
maybe delete this since there could be nulls when transaction is pending

chuxinh commented on models/core/txs.yml:
```diff
@@ -0,0 +1,82 @@
+
+      - name: block_number
+        description: "Block number where this transaction was in. `null` when it's pending."
+        tests:
+          - not_null
```
same for here

chuxinh commented on models/core/txs.sql:
```diff
@@ -1,170 +1,25 @@
--- or https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models
+    tx:gas as gas,
+    tx_id as tx_hash,
+    tx :input::string as data,
```
```suggestion
    tx:input::string as data,
```

chuxinh reviewed (changes requested):
nits. 
the logics looks good to me. and try rerunning the dbt in dev again to make sure it's running. 😄 

also feel free if you want to adopt the format of using CTEs like the one for blocks https://github.com/MetricsDAO/harmony_dbt/pull/10/files
this is suggested by @forgxyz with the guide from here: https://discourse.getdbt.com/t/why-the-fishtown-sql-style-guide-uses-so-many-ctes/1091

chuxinh commented:
@mattromano i just merged `blocks` into prod so no need to delete `schema.yml` in the PR.
You might want to do a git rebase for the branch

mattromano committed:
rem

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
rem

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
updated tx_hash name in config

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
updated tx_hash name in config

Co-authored-by: Chuxin Huang <30974572+chuxinh@users.noreply.github.com>

mattromano committed:
removed not null tests for block_hash and tx_hash

mattromano committed:
removed not null tests for block_hash and tx_hash

mattromano committed:
remove quotes and update table desc

mattromano committed:
remove quotes and update table desc

mattromano committed:
added CTE formatting

mattromano committed:
added CTE formatting

mattromano committed:
update dedupe to correct macro name

mattromano committed:
update dedupe to correct macro name

mattromano committed:
Merge branch 'main' of github.com:MetricsDAO/harmony_dbt

mattromano committed:
Merge branch 'add-txn' of github.com:MetricsDAO/harmony_dbt into add-txn

mattromano committed:
update to match final tx_hash header name

mattromano commented:
### Updates
- YML removal of quotes + dropped not-null tests for block hash and block number
- Updated query to recommended CTE formatting 
- Rebased with main branch
 
Also I was able to do a full dbt run of the newly rebased branch:
```
root@4c27223889b8:/harmony#dbt run
Running with dbt=0.21.1
Found 3 models, 31 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

22:15:40 | Concurrency: 4 threads (target='dev')
22:15:40 | 
22:15:40 | 1 of 3 START incremental model DEV.blocks............................ [RUN]
22:15:40 | 2 of 3 START incremental model DEV.txs............................... [RUN]
22:15:51 | 1 of 3 OK created incremental model DEV.blocks....................... [SUCCESS 1 in 10.63s]
22:15:51 | 3 of 3 START table model DEV.metric_miner_count...................... [RUN]
22:15:54 | 3 of 3 OK created table model DEV.metric_miner_count................. [SUCCESS 1 in 3.08s]
22:21:25 | 2 of 3 OK created incremental model DEV.txs.......................... [SUCCESS 1 in 344.29s]
22:21:25 | 
22:21:25 | Finished running 2 incremental models, 1 table model in 347.04s.

Completed successfully
```

Also ran `dbt test` and everything checks out there as well.

```
root@4c27223889b8:/harmony# dbt test
Running with dbt=0.21.1
Found 3 models, 31 tests, 0 snapshots, 0 analyses, 175 macros, 0 operations, 0 seed files, 2 sources, 0 exposures

22:24:08 | Concurrency: 4 threads (target='dev')
22:24:08 | 
22:24:08 | 1 of 31 START test not_null_blocks_block_hash........................ [RUN]
22:24:08 | 2 of 31 START test not_null_blocks_block_id.......................... [RUN]
22:24:08 | 3 of 31 START test not_null_blocks_block_parent_hash................. [RUN]
22:24:08 | 4 of 31 START test not_null_blocks_block_timestamp................... [RUN]
22:24:09 | 3 of 31 PASS not_null_blocks_block_parent_hash....................... [PASS in 1.59s]
22:24:09 | 5 of 31 START test not_null_blocks_gas_limit......................... [RUN]
22:24:10 | 4 of 31 PASS not_null_blocks_block_timestamp......................... [PASS in 1.78s]
22:24:10 | 6 of 31 START test not_null_blocks_gas_used.......................... [RUN]
22:24:10 | 1 of 31 PASS not_null_blocks_block_hash.............................. [PASS in 1.97s]
22:24:10 | 7 of 31 START test not_null_blocks_miner............................. [RUN]
22:24:10 | 2 of 31 PASS not_null_blocks_block_id................................ [PASS in 2.22s]
22:24:10 | 8 of 31 START test not_null_blocks_nonce............................. [RUN]
22:24:11 | 6 of 31 PASS not_null_blocks_gas_used................................ [PASS in 1.50s]
22:24:11 | 9 of 31 START test not_null_blocks_receipts_root..................... [RUN]
22:24:11 | 5 of 31 PASS not_null_blocks_gas_limit............................... [PASS in 1.76s]
22:24:11 | 10 of 31 START test not_null_blocks_size............................. [RUN]
22:24:11 | 8 of 31 PASS not_null_blocks_nonce................................... [PASS in 1.35s]
22:24:11 | 11 of 31 START test not_null_blocks_state_root....................... [RUN]
22:24:12 | 7 of 31 PASS not_null_blocks_miner................................... [PASS in 1.67s]
22:24:12 | 12 of 31 START test not_null_blocks_tx_count......................... [RUN]
22:24:13 | 9 of 31 PASS not_null_blocks_receipts_root........................... [PASS in 1.54s]
22:24:13 | 13 of 31 START test not_null_metric_miner_count_metric_date.......... [RUN]
22:24:13 | 10 of 31 PASS not_null_blocks_size................................... [PASS in 1.61s]
22:24:13 | 14 of 31 START test not_null_metric_miner_count_metric_period........ [RUN]
22:24:13 | 11 of 31 PASS not_null_blocks_state_root............................. [PASS in 1.70s]
22:24:13 | 12 of 31 PASS not_null_blocks_tx_count............................... [PASS in 1.49s]
22:24:13 | 15 of 31 START test not_null_metric_miner_count_miner_count.......... [RUN]
22:24:13 | 16 of 31 START test not_null_txs_block_timestamp..................... [RUN]
22:24:14 | 13 of 31 PASS not_null_metric_miner_count_metric_date................ [PASS in 1.69s]
22:24:14 | 17 of 31 START test not_null_txs_data................................ [RUN]
22:24:14 | 14 of 31 PASS not_null_metric_miner_count_metric_period.............. [PASS in 1.55s]
22:24:14 | 18 of 31 START test not_null_txs_from_address........................ [RUN]
22:24:15 | 16 of 31 PASS not_null_txs_block_timestamp........................... [PASS in 1.51s]
22:24:15 | 19 of 31 START test not_null_txs_gas................................. [RUN]
22:24:15 | 15 of 31 PASS not_null_metric_miner_count_miner_count................ [PASS in 1.74s]
22:24:15 | 20 of 31 START test not_null_txs_gas_price........................... [RUN]
22:24:16 | 17 of 31 PASS not_null_txs_data...................................... [PASS in 1.51s]
22:24:16 | 21 of 31 START test not_null_txs_native_from_address................. [RUN]
22:24:16 | 18 of 31 PASS not_null_txs_from_address.............................. [PASS in 1.57s]
22:24:16 | 22 of 31 START test not_null_txs_native_to_address................... [RUN]
22:24:17 | 19 of 31 PASS not_null_txs_gas....................................... [PASS in 2.24s]
22:24:17 | 23 of 31 START test not_null_txs_nonce............................... [RUN]
22:24:17 | 20 of 31 PASS not_null_txs_gas_price................................. [PASS in 2.11s]
22:24:17 | 24 of 31 START test not_null_txs_status.............................. [RUN]
22:24:17 | 21 of 31 PASS not_null_txs_native_from_address....................... [PASS in 1.46s]
22:24:17 | 25 of 31 START test not_null_txs_to_address.......................... [RUN]
22:24:17 | 22 of 31 PASS not_null_txs_native_to_address......................... [PASS in 1.39s]
22:24:17 | 26 of 31 START test not_null_txs_tx_hash............................. [RUN]
22:24:19 | 24 of 31 PASS not_null_txs_status.................................... [PASS in 1.46s]
22:24:19 | 27 of 31 START test not_null_txs_value............................... [RUN]
22:24:19 | 23 of 31 PASS not_null_txs_nonce..................................... [PASS in 1.81s]
22:24:19 | 28 of 31 START test unique_blocks_block_hash......................... [RUN]
22:24:19 | 25 of 31 PASS not_null_txs_to_address................................ [PASS in 1.61s]
22:24:19 | 26 of 31 PASS not_null_txs_tx_hash................................... [PASS in 1.62s]
22:24:19 | 29 of 31 START test unique_blocks_block_id........................... [RUN]
22:24:19 | 30 of 31 START test unique_blocks_block_parent_hash.................. [RUN]
22:24:20 | 27 of 31 PASS not_null_txs_value..................................... [PASS in 1.78s]
22:24:20 | 31 of 31 START test unique_txs_tx_hash............................... [RUN]
22:24:21 | 29 of 31 PASS unique_blocks_block_id................................. [PASS in 2.01s]
22:24:21 | 28 of 31 PASS unique_blocks_block_hash............................... [PASS in 2.19s]
22:24:21 | 30 of 31 PASS unique_blocks_block_parent_hash........................ [PASS in 2.31s]
22:24:28 | 31 of 31 PASS unique_txs_tx_hash..................................... [PASS in 7.90s]
22:24:28 | 
22:24:28 | Finished running 31 tests in 21.88s.
```

chuxinh reviewed (approved):
Thanks for all the changes made! Looks good to me!
